design/history/exploration-reports/2021.02-cross-language-fixtures.md

```diff
@@ -0,0 +1,141 @@
+	- e.g.: "put cursor at end of file, press enter, press dash three times" -- something like this complexity level is the goal.
+- should preserve whitespace precisely.
+	- it would be problematic for parser fixtures if whitespace isn't preserved!  (Mind touchy things like trailing linebreaks, especially.)
+- must be able to frame incomplete sub-documents, and not require knowledge of the parser for them.
+	- I need to use this to write fixtures for parsers _and their error modes_, which means sub-documents might be _intentionally_ unparsable.
+- nonrequirement: able to represent binary.
+	- as close as possible is certainly better.
+	- this can't be a goal simply because it runs directly crosswise to "easy to edit as a human".
+	- we may want to make a convention around base64 (?) and possible around comments in that format (?) ... but this can be implemented separately from the overall document container format.
+- nonrequirement: recursiveness.
+	- in practice, test fixtures tend not to require this.
+	- less complexity is better.
+- nonrequirement: anything like "BDD" or smelling like cucumber.
+	- that's just not the goal here.
+- stretch goal: programmatically updateable.  In other words: I should be able to write half a fixture, run a program, and have it write the rest of the file.
+	- if this is supported, ideally we'll have a check in CI that makes sure all files are canonically formatted.
+- stretch goal: handles content-addressable documents well.
+	- this is a stretch goal rather than a hard requirement because: content-addressable DAGs are _common_ in our work, but not _universal_.  Some things we write fixtures for don't involve DAGs, so we should solve that problem first, and then figure out how to also handle DAGs.
+	- this is a stretch goal rather than a hard requirement because: this author has the hypothesis this will be easy to add to existing container formats.
+
+
+prior art
+---------
+
+- selector fixtures in json: https://github.com/ChainSafe/ipld-traversal-vectors
+	- notes on outcome: awesome... except when people eyeball it, the fact that the top level maps are containing "control plane" info (which part of the fixture it is) is non-obvious.
+- schema DSL parser fixtures: https://github.com/ipld/js-ipld-schema/tree/master/test/fixtures and https://github.com/ipld/go-ipld-schema/tree/master/test/fixtures
+	- notes on outcome: the parts that use yaml: awesome... although a lot of the yaml files are resorting to heredocs.
+		- heredocs aren't always preserved by yaml parse and reserialize, so that's a knock against the "programatically updatable" goal.
+		- This author is frankly just not a fan of yaml.  The parsers are... complex.
+	- notes on outcome: the parts that don't use yaml: awesome... except sometimes there's two files that are siblings, and they're related, and you just have to "know" that.
+- (other project) "interactive" fixtures in noise: https://github.com/pipedown/noise/blob/master/repl-tests/limit.noise
```

Comment:
The prior art to that one is https://dev.mysql.com/doc/dev/mysql-server/latest/PAGE_MYSQL_TEST_RUN.html

Reply:
Added entry.  :+1: